crates/rust-analyzer/src/conv.rs

```diff
@@ -163,6 +163,14 @@ impl ConvWith<(&LineIndex, LineEndings)> for CompletionItem {
             ..Default::default()
         };
 
+        if let Some(score) = self.score() {
+            match score {
+                CompletionScore::TypeAndNameMatch => res.preselect = Some(true),
+                CompletionScore::TypeMatch => {}
+            }
+            res.sort_text = Some(format!("{:02}", ctx.2));
```

Comment:
I'd rather start with just setting preselect and void the sort_text hack. I belive setitng sort_text only for some fields would be wrong, and we also need to take fuzzy mathcing into account, if we set sort text. I think just preselect would give a meaningful improvement.

Reply:
If you have an example in mind in which case sort text could cause an issue let me know I will try. The fact is as a rust analyzer user I feel better with my example in sort text. But of course I don’t want to introduce a bug or non deterministic behavior. So if you have examples on which I can make some tests I will be happy to try and show you the result. I agree with you, preselect is good but if we can’t go a step further using sortText as gopls does. Developer experience will be improved.

Reply:
By the way when I go to the LSP specs from Microsoft, when I look about documentation about `sortText` here it is `A string that should be used when comparing this item with other items. When falsy the label is used.` it seems it's not a hack. I mean, today we have all completion items sorted in alphabetical, which in my opinion, is very bad. IMHO I think we can't do something more bad because you always have to scroll the entire autocompletions list to see all interesting parameters of a struct for example. I think it really worth to at least show to the user fields with right types first + preselect perfect match. If we look at my screenshots there are keywords like `box`, `dbg`, ... displayed and at this place we don't care about it. In my example it's not so annoying because I have a lot of fields beginning with `a` letter so it's in first position and I don't have to scroll so much. If you are worried I can write a lot of tests, make a lot of experiments even manually and show you the results because I really think it worth.